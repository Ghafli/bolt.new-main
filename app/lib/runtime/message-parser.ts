import { Action } from "@/app/types/actions";
import { stripIndent } from "@/app/utils/stripIndent";

export class ActionRunnerError extends Error {}

export function parseCommands(input: string): Action[] | null {
  const trimmed = input.trim();

  if (!trimmed) {
    return null;
  }

  const lines = trimmed.split("\n");

  const actions: Action[] = [];

  for (const line of lines) {
    const trimmedLine = line.trim();
    const parts = trimmedLine.split(" ");
    const command = parts[0];
    const args = parts.slice(1);

    if (!command) {
      continue;
    }


    actions.push({
      command,
      arguments: args,
    });
  }

  return actions.length > 0 ? actions : null;
}

export function parseCodeBlock(input: string): string | null {
  const trimmed = stripIndent(input).trim();

  if (!trimmed) {
    return null;
  }

  const codeBlockRegex = /```.*\n([\s\S]*?)```/g;
  const match = codeBlockRegex.exec(trimmed);

    if (!match) {
      return null;
    }

  return match[1]
}
