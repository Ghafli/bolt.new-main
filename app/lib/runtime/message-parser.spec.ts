import { parseCommands, ActionRunnerError } from "./message-parser";

describe("message parser", () => {
  describe("command parsing", () => {
    it("should parse simple command", () => {
      expect(parseCommands("ls")).toMatchSnapshot();
    });

    it("should parse simple command with args", () => {
      expect(parseCommands("npm install react")).toMatchSnapshot();
    });

    it("should parse commands with arguments", () => {
      expect(parseCommands("run npm install react")).toMatchSnapshot();
    });

    it("should parse commands with string argument", () => {
      expect(parseCommands("cd /some/path/here")).toMatchSnapshot();
    });

    it("should parse commands with no arguments", () => {
      expect(parseCommands("ls")).toMatchSnapshot();
    });


    it("should return null if no command", () => {
      expect(parseCommands("hello world")).toMatchSnapshot();
    });

    it("should parse multiple commands with arguments", () => {
      expect(parseCommands("run npm install react\nls\ncd /some/path/here")).toMatchSnapshot();
    });

    it("should parse multiple commands with no arguments", () => {
       expect(parseCommands("ls\npwd")).toMatchSnapshot();
    });

     it("should trim input", () => {
      expect(parseCommands("  ls  ")).toMatchSnapshot();
    });

  });


});
