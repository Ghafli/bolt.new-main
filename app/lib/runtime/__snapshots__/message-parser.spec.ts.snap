// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message parser command parsing should parse commands with arguments 1`] = `
{
  "command": "run",
  "arguments": [
    "npm",
    "install",
    "react",
  ],
}
`;

exports[`message parser command parsing should parse commands with no arguments 1`] = `
{
  "command": "ls",
  "arguments": [],
}
`;

exports[`message parser command parsing should parse commands with string argument 1`] = `
{
  "command": "cd",
  "arguments": [
    "/some/path/here",
  ],
}
`;

exports[`message parser command parsing should parse multiple commands with arguments 1`] = `
[
  {
    "command": "run",
    "arguments": [
      "npm",
      "install",
      "react",
    ],
  },
  {
    "command": "ls",
    "arguments": [],
  },
  {
    "command": "cd",
    "arguments": [
      "/some/path/here",
    ],
  },
]
`;

exports[`message parser command parsing should parse multiple commands with no arguments 1`] = `
[
  {
    "command": "ls",
    "arguments": [],
  },
  {
    "command": "pwd",
    "arguments": [],
  },
]
`;

exports[`message parser command parsing should parse simple command 1`] = `
{
  "command": "ls",
  "arguments": [],
}
`;

exports[`message parser command parsing should parse simple command with args 1`] = `
{
  "command": "npm",
  "arguments": [
    "install",
    "react",
  ],
}
`;

exports[`message parser command parsing should return null if no command 1`] = `null`;

exports[`message parser command parsing should trim input 1`] = `
{
  "command": "ls",
  "arguments": [],
}
`;
